test/package/test_torchscript.py

```diff
@@ -0,0 +1,584 @@
+        scripted_mod_bar = torch.jit.script(ModBar("bar"))
+
+        buffer_0 = BytesIO()
+        with PackageExporter(buffer_0, verbose=False) as e:
+            e.save_pickle("res", "mod1.pkl", scripted_mod_foo)
+
+        buffer_0.seek(0)
+        importer_0 = importer = PackageImporter(buffer_0)
+
+        buffer_1 = BytesIO()
+        with PackageExporter(buffer_1, verbose=False) as e:
+            e.save_pickle("res", "mod1.pkl", scripted_mod_bar)
+
+        buffer_1.seek(0)
+        importer_1 = PackageImporter(buffer_1)
+
+        self.assertTrue("torchvision" in str(importer_0.file_structure()))
+        self.assertFalse("torchvision" in str(importer_1.file_structure()))
+
+    def test_save_scriptmodules_in_container(self):
+        """
+        Test saving of ScriptModules inside of container.
+        """
+
+        class ModB(torch.nn.Module):
+            def __init__(self, name: str):
+                super().__init__()
+                self.name = name
+
+            def forward(self, input: str):
+                input = input + "_modB:" + self.name
+                return input
+
+        class ModA(torch.nn.Module):
+            def __init__(self, name: str):
+                super().__init__()
+                self.name = name
+
+            def forward(self, input: str):
+                input = input + "_modA:" + self.name
+                return input
+
+        scripted_mod_a = torch.jit.script(ModA("a"))
+        scripted_mod_b = torch.jit.script(ModB("b"))
+        script_mods_list = [scripted_mod_a, scripted_mod_b]
+
+        buffer = BytesIO()
+        with PackageExporter(buffer, verbose=False) as e:
+            e.save_pickle("res", "list.pkl", script_mods_list)
+
+        buffer.seek(0)
+        importer = PackageImporter(buffer)
+        loaded_mod_list = importer.load_pickle("res", "list.pkl")
+        self.assertEqual(loaded_mod_list[0]("input"), scripted_mod_a("input"))
+        self.assertEqual(loaded_mod_list[1]("input"), scripted_mod_b("input"))
+
+    def test_save_shared_scriptmodules(self):
+        """
+        Test saving of single ScriptModule shared by multiple
+        eager modules (ScriptModule should be saved just once).
+        """
+
+        from package_a.test_module import SimpleTest, Mod
+
+        scripted_mod = torch.jit.script(SimpleTest())
+
+        mod1 = Mod(scripted_mod)
+        mod2 = Mod(scripted_mod)
+        mod3 = Mod(scripted_mod)
+
+        buffer = BytesIO()
+        with PackageExporter(buffer, verbose=False) as e:
+            e.save_pickle("res", "mod1.pkl", mod1)
+            e.save_pickle("res", "mod2.pkl", mod2)
+            e.save_pickle("res", "mod3.pkl", mod3)
+
+        buffer.seek(0)
+        importer = PackageImporter(buffer)
+        file_structure = importer.file_structure()
+        self.assertTrue(file_structure.has_file(".data/ts_code/0"))
+        self.assertFalse(file_structure.has_file(".data/ts_code/1"))
```

Comment:
We want to check two things:
1. On save, we only serialize once
2. On load, we only deserialize once (e.g. we share the same ScriptModule object across all three loaded modules)

Right now we're checking 1), but let's check 2) as well. We can add an identity assertion: if we load_pickle the three mods, the `id` of their script_mod attribute should be the same, right? Or maybe the Python wrapper is different, but the underlying C impl should be the same.

Reply:
I think I'm confused about how pickled object inside of a package are supposed to relate to each other. Are the relationships between objects supposed to be upheld between `load_pickle`s? I can understand preserving the relationships between objects that are contained in a parent object that is loaded from a pickle, but not between `load_pickle`s. The current `PackageImporter.load_pickle` reads to me like it only preserves the relationship between storages within a `load_pickle`-- in other words each `load_pickle` will re-load the storages newly each time. We could implement a cache on the importer to allow storages to be loaded only once, and for script-modules, but idk if that matches the normal pickle behavior. If a user pickle_loads the same pickle twice, should they get the same object by python `id` twice? We don't do that currently

Reply:
The real thing I'm confused about is if `load_pickle` is supposed to return a new instance of the saved object each time, or if it's always supposed to return the same instance. Like if you call `load_pickle("res", "mod1.pkl")` twice, should both returned objects be the same instance? Right now they're not:
```
''' Mod definition:
class Mod(torch.nn.Module):
    def __init__(self, x, y):
        super().__init__()
        self._list = x
        self._tensor = y

    def forward(self, x):
        return 
'''

mod = Mod([], torch.rand(1,2,3))

buffer = BytesIO()
with PackageExporter(buffer, verbose=False) as e:
    e.extern(include=["torch", "sys", "io"])
    e.save_pickle("model", "model.pkl", mod)

buffer.seek(0)
importer = PackageImporter(buffer)

loaded_mod1 = importer.load_pickle("model", "model.pkl")
loaded_mod1_repeat = importer.load_pickle("model", "model.pkl") 

assert id(loaded_mod1) != id(loaded_mod1_repeat)
assert id(loaded_mod1._tensor) != id(loaded_mod1_repeat._tensor)

assert torch.allclose(loaded_mod1._tensor, loaded_mod1_repeat._tensor) == 1
loaded_mod1._tensor = loaded_mod1._tensor * 2
assert torch.allclose(loaded_mod1._tensor, loaded_mod1_repeat._tensor) == 0
```

Reply:
Ah sorry I think my suggestion was wrong. You're right: each call to `load_pickle` should produce a new copy of everything.

I was thinking of like a diamond shape:
```
        parent
       /  |  \
      A   B   C
      \   |   /
       child
```

if you do `load_pickle(parent)`, it should be the case that `a.A.child is a.B.child is a.C.child`, where child can either be a tensor or a scriptmodule.

Reply:
Awesome. The code now preserves this relationship on load for both ScriptModules and storages. The tests to show them have '_load' in the name of the test